File: CHAPTER 6. 자바에서 코틀린 컬렉션으로/yckim/chapter6/java/v1/Suffering.java
package chapter6.java.v1;

import chapter6.java.Journey;
import chapter6.java.Location;

import java.util.List;

import static chapter6.java.Other.SOME_COMPLICATED_RESULT;
import static chapter6.java.Other.routesFor;
import static chapter6.java.Route.getDepartsFrom;
import static java.util.Comparator.comparing;

public class Suffering {
    public static int sufferScoreFor(List<Journey> route) {
        List<Journey> longestJourneys = longestJourneysIn(route, 3);
        return sufferScore(longestJourneys, getDepartsFrom(route));
    }

    public static List<Journey> longestJourneysIn(
            List<Journey> journeys,
            int limit
    ) {
        journeys.sort(comparing(Journey::getDuration).reversed()); // <1>
        var actualLimit = Math.min(journeys.size(), limit);
        return journeys.subList(0, actualLimit);
    }

    public static List<List<Journey>> routesToShowFor(String itineraryId) {
        var routes = routesFor(itineraryId);
        removeUnbearableRoutes(routes);
        return routes;
    }

    private static void removeUnbearableRoutes(List<List<Journey>> routes) {
        routes.removeIf(route -> sufferScoreFor(route) > 10);
    }

    private static int sufferScore(
            List<Journey> longestJourneys,
            Location start
    ) {
        return SOME_COMPLICATED_RESULT();
    }
}

File: CHAPTER 6. 자바에서 코틀린 컬렉션으로/yckim/chapter6/kotlin/v1/Suffering.kt
package chapter6.kotlin.v1

import chapter6.java.Journey
import chapter6.java.Location
import chapter6.java.Other
import chapter6.java.Route
import java.util.*

object Suffering {
    @JvmStatic
    fun sufferScoreFor(route: List<Journey>): Int {
        return sufferScore(
            route.longestJourneys(3),
            Route.getDepartsFrom(route)
        )
    }

    @JvmStatic
    fun List<Journey>.longestJourneys(
        limit: Int
    ): List<Journey> = sortedByDescending { it.duration }.take(limit)

    fun routesToShowFor(itineraryId: String?): List<List<Journey>> {
        return bearable(Other.routesFor(itineraryId))
    }

    private fun bearable(routes: List<List<Journey>>): List<List<Journey>> =
        routes.filter { sufferScoreFor(it) <= 10 }

    private fun sufferScore(
        longestJourneys: List<Journey>,
        start: Location
    ): Int {
        return Other.SOME_COMPLICATED_RESULT()
    }

    inline fun <reified E> sorted(
        collection: Collection<E>,
        by: Comparator<in E>?
    ): List<E> {
        val result = collection.toTypedArray()
        Arrays.sort(result, by)
        return Arrays.asList(*result)
    }
}